File: app/src/main/java/com/example/match3game/BoardView.java
package com.example.match3game;

import android.content.Context;
import android.graphics.*;
import android.view.MotionEvent;
import android.view.View;
import android.util.AttributeSet;
import java.util.ArrayList;
import java.util.List;

/**
 * 棋盘视图
 * 自定义 View 用于绘制游戏棋盘
 */
public class BoardView extends View {
    private static final int BOARD_SIZE = 6;
    private static final int GEM_COLORS = 7;

    private GameEngine gameEngine;
    private AdManager adManager;

    private float gemSize;
    private float padding;
    private float boardPadding;

    private int[] colors = new int[]{
            0xFFFF6B6B,   // 红色
            0xFF4ECDC4,   // 青色
            0xFF45B7D1,   // 蓝色
            0xFF96CEB4,   // 绿色
            0xFFE6D08D,   // 黄色
            0xFFD4A5A5,   // 紫色
            0xFF98D8C8    // 粉色
    };

    private Gem selectedGem;
    private Point selectedGemPosition;

    public BoardView(Context context) {
        super(context);
        init(context);
    }

    public BoardView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        gameEngine = new GameEngine();
        selectedGem = null;
        selectedGemPosition = new Point();

        setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_DOWN) {
                    float x = event.getX();
                    float y = event.getY();
                    handleGemClick(x, y);
                }
                return true;
            }
        });

        // 初始化广告
        adManager = new AdManager((MainActivity) context);

        // 资源尺寸
        float screenWidth = context.getResources().getDisplayMetrics().widthPixels;
        float screenHeight = context.getResources().getDisplayMetrics().heightPixels;

        // 计算宝石大小
        float availableWidth = screenWidth - 40;
        float availableHeight = screenHeight - 200;
        gemSize = Math.min(availableWidth / BOARD_SIZE, availableHeight / BOARD_SIZE);

        padding = (screenWidth - gemSize * BOARD_SIZE) / 2;
        boardPadding = 10f;
    }

    /**
     * 处理宝石点击
     */
    private void handleGemClick(float x, float y) {
        int col = (int) ((x - padding - boardPadding) / gemSize);
        int row = (int) ((y - padding - boardPadding) / gemSize);

        if (row >= 0 && row < BOARD_SIZE && col >= 0 && col < BOARD_SIZE) {
            boolean success = gameEngine.handleGemClick(row, col);
            if (success) {
                invalidate();
                checkForMatches();
            }
        }
    }

    /**
     * 检查是否有匹配
     */
    private void checkForMatches() {
        List<List<Gem>> matches = gameEngine.findMatches();
        if (!matches.isEmpty()) {
            postDelayed(new Runnable() {
                @Override
                public void run() {
                    gameEngine.removeMatches();
                    invalidate();

                    if (gameEngine.isGameOver()) {
                    }
                }
            }, 300);
        }
    }

    /**
     * 绘制宝石
     */
    private void drawGem(Canvas canvas, Gem gem, float x, float y) {
        if (gem == null) return;

        Paint paint = new Paint();
        paint.setColor(colors[gem.color]);
        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);

        RectF rect = new RectF(
                x + boardPadding,
                y + boardPadding,
                x + gemSize - boardPadding,
                y + gemSize - boardPadding
        );
        canvas.drawRoundRect(rect, 10f, 10f, paint);

        // 绘制边框
        if (gem.isSelected) {
            Paint strokePaint = new Paint();
            strokePaint.setColor(Color.WHITE);
            strokePaint.setStyle(Paint.Style.STROKE);
            strokePaint.setStrokeWidth(5f);
            strokePaint.setAntiAlias(true);
            canvas.drawRoundRect(rect, 10f, 10f, strokePaint);
        }

        // 绘制高光
        Paint highlightPaint = new Paint();
        highlightPaint.setColor(Color.WHITE);
        highlightPaint.setAlpha(100);
        highlightPaint.setStyle(Paint.Style.FILL);
        highlightPaint.setAntiAlias(true);
        canvas.drawRoundRect(
                rect.left + 5,
                rect.top + 5,
                rect.right - 10,
                rect.bottom - 10,
                8f,
                8f,
                highlightPaint
        );
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.drawColor(Color.parseColor("#2C3E50"));

        RectF boardRect = new RectF(
                padding,
                padding,
                getWidth() - padding,
                getHeight() - padding
        );
        Paint boardPaint = new Paint();
        boardPaint.setColor(Color.parseColor("#34495E"));
        boardPaint.setStyle(Paint.Style.FILL);
        boardPaint.setAntiAlias(true);
        canvas.drawRoundRect(boardRect, 15f, 15f, boardPaint);

        for (int row = 0; row < BOARD_SIZE; row++) {
            for (int col = 0; col < BOARD_SIZE; col++) {
                Gem gem = gameEngine.getGemAt(row, col);
                float x = padding + col * gemSize;
                float y = padding + row * gemSize;
                drawGem(canvas, gem, x, y);
            }
        }
    }

    /**
     * 重置游戏
     */
    public void resetGame() {
        gameEngine = new GameEngine();
        invalidate();
    }

    /**
     * 获取当前分数
     */
    public int getScore() {
        return gameEngine.getScore();
    }

    /**
     * 获取剩余步数
     */
    public int getMoves() {
        return gameEngine.getMoves();
    }
}

File: app/src/main/java/com/example/match3game/BoardView.kt
package com.example.match3game

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.core.content.res.ResourcesCompat
import kotlin.math.abs

/**
 * 棋盘视图
 * 自定义 View 用于绘制游戏棋盘
 */
class BoardView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val gameEngine = GameEngine()
    private val adManager = AdManager(context as MainActivity) // TODO: 传入正确的 Activity

    private val gemSize: Float
    private val padding: Float
    private val boardPadding: Float
    private val colors = listOf(
        Color.parseColor("#FF6B6B"),   // 红色
        Color.parseColor("#4ECDC4"),   // 青色
        Color.parseColor("#45B7D1"),   // 蓝色
        Color.parseColor("#96CEB4"),   // 绿色
        Color.parseColor("#FFEAA7"),   // 黄色
        Color.parseColor("#DDA0DD"),   // 紫色
        Color.parseColor("#98D8C8")    // 粉色
    )

    private val selectedGem: Gem? = null // 临时存储选中的宝石
    private val selectedGemPosition: Point = Point()

    init {
        // 设置点击监听
        setOnTouchListener { view, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    val x = event.x
                    val y = event.y
                    handleGemClick(x, y)
                }
            }
            true
        }

        // 初始化广告
        adManager.initialize()

        // 资源尺寸
        val displayMetrics = resources.displayMetrics
        val screenWidth = displayMetrics.widthPixels
        val screenHeight = displayMetrics.heightPixels

        // 计算宝石大小（留出边距）
        val availableWidth = screenWidth - 40
        val availableHeight = screenHeight - 200
        gemSize = minOf(availableWidth / 6, availableHeight / 6)

        padding = (screenWidth - gemSize * 6) / 2
        boardPadding = 10f
    }

    /**
     * 处理宝石点击
     */
    private fun handleGemClick(x: Float, y: Float) {
        // 计算点击的行列
        val col = ((x - padding - boardPadding) / gemSize).toInt()
        val row = ((y - padding - boardPadding) / gemSize).toInt()

        if (row >= 0 && row < 6 && col >= 0 && col < 6) {
            val success = gameEngine.handleGemClick(row, col)
            if (success) {
                invalidate() // 重绘棋盘
                checkForMatches()
            }
        }
    }

    /**
     * 检查是否有匹配
     */
    private fun checkForMatches() {
        val matches = gameEngine.findMatches()
        if (matches.isNotEmpty()) {
            // 延迟执行，让动画播放
            postDelayed({
                gameEngine.removeMatches()
                invalidate()

                // 更新分数
                val score = gameEngine.getScore()
                // TODO: 更新 UI 显示分数

                // 检查游戏是否结束
                if (gameEngine.isGameOver()) {
                    // 游戏结束，显示插屏广告
                    adManager.showAdOnGameOver()
                }
            }, 300)
        }
    }

    /**
     * 绘制宝石
     */
    private fun drawGem(canvas: Canvas, gem: Gem?, x: Float, y: Float) {
        if (gem == null) return

        val paint = Paint().apply {
            color = colors[gem.color]
            style = Paint.Style.FILL
            isAntiAlias = true
        }

        // 绘制圆角矩形
        val rect = RectF(
            x + boardPadding,
            y + boardPadding,
            x + gemSize - boardPadding,
            y + gemSize - boardPadding
        )
        canvas.drawRoundRect(rect, 10f, 10f, paint)

        // 如果被选中，绘制边框
        if (gem.isSelected) {
            val strokePaint = Paint().apply {
                color = Color.WHITE
                style = Paint.Style.STROKE
                strokeWidth = 5f
                isAntiAlias = true
            }
            canvas.drawRoundRect(rect, 10f, 10f, strokePaint)
        }

        // 绘制高光效果
        val highlightPaint = Paint().apply {
            color = Color.parseColor("#FFFFFF")
            alpha = 100
            style = Paint.Style.FILL
            isAntiAlias = true
        }
        canvas.drawRoundRect(
            rect.left + 5,
            rect.top + 5,
            rect.right - 10,
            rect.bottom - 10,
            8f,
            8f,
            highlightPaint
        )
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // 清空画布
        canvas.drawColor(Color.parseColor("#2C3E50"))

        // 绘制棋盘背景
        val boardRect = RectF(
            padding,
            padding,
            width - padding,
            height - padding
        )
        val boardPaint = Paint().apply {
            color = Color.parseColor("#34495E")
            style = Paint.Style.FILL
            isAntiAlias = true
        }
        canvas.drawRoundRect(boardRect, 15f, 15f, boardPaint)

        // 绘制每个宝石
        for (row in 0 until 6) {
            for (col in 0 until 6) {
                val gem = gameEngine.getGemAt(row, col)
                val x = padding + col * gemSize
                val y = padding + row * gemSize
                drawGem(canvas, gem, x, y)
            }
        }
    }

    /**
     * 重置游戏
     */
    fun resetGame() {
        gameEngine = GameEngine()
        invalidate()
    }

    /**
     * 获取当前分数
     */
    fun getScore(): Int {
        return gameEngine.getScore()
    }

    /**
     * 获取剩余步数
     */
    fun getMoves(): Int {
        return gameEngine.getMoves()
    }
}
